### DISCUSSION TITLE
"meepctl dockerize" command failing

### DISCUSSION THREAD

**Discussion by anokun7** [InterDigitalInc/AdvantEDGE#183] [Q&A] [answered]
The `meepctl dockerize` commands are failing in a fresh setup.

Below are the commands I tried.
```
[ec2-user@ip-172-31-25-121 ~]$ git clone https://github.com/InterDigitalInc/AdvantEDGE.git
Cloning into 'AdvantEDGE'...
<snip>
Resolving deltas: 100% (19356/19356), done.
```
```
[ec2-user@ip-172-31-25-121 ~]$ meepctl config
Using repo config file: /home/ec2-user/AdvantEDGE/.meepctl-repocfg.yaml
Using meepctl config file: /home/ec2-user/.meepctl.yaml
Get/Set meep environment configuration.
Config file location: ~/.meepctl.yaml
<snip>
========================================
meep.gitdir : /home/ec2-user/AdvantEDGE
meep.registry : anoop
meep.workdir : /home/ec2-user/.meep
node.ip : 172.31.25.121
========================================
```
```
[ec2-user@ip-172-31-25-121 ~]$ meepctl dockerize all -v
Using repo config file: /home/ec2-user/AdvantEDGE/.meepctl-repocfg.yaml
Using meepctl config file: /home/ec2-user/.meepctl.yaml
Dockerize called
[arg]  targets: [all]
[flag] registry:
[flag] verbose: true
[flag] time: false
Using docker registry: anoop
-- meep-mon-engine --
   + copy docker data
Cmd: [/bin/sh -c find /home/ec2-user/AdvantEDGE/bin/meep-mon-engine -type f | xargs sha256sum | sort | sha256sum]
Result: find: ‘/home/ec2-user/AdvantEDGE/bin/meep-mon-engine’: No such file or directory
abcfa6a9d4df344d1781bc2560b5e4cdcae08b39ed303063535e7e1e926a304a  -

   + dockerize meep-mon-engine
Cmd: [docker build --no-cache --rm --label MeepVersion=find: -t anoop/meep-mon-engine /home/ec2-user/AdvantEDGE/bin/meep-mon-engine]
Result: unable to prepare context: path "/home/ec2-user/AdvantEDGE/bin/meep-mon-engine" not found

Error: Failed to dockerize  anoop/meep-mon-engine  with error:  exit status 1
<snip>
```

Building for instance, just the `meep-virt-engine` image
```
[ec2-user@ip-172-31-25-121 ~]$ meepctl dockerize meep-virt-engine -v
Using repo config file: /home/ec2-user/AdvantEDGE/.meepctl-repocfg.yaml
Using meepctl config file: /home/ec2-user/.meepctl.yaml
Dockerize called
[arg]  targets: [meep-virt-engine]
[flag] registry:
[flag] verbose: true
[flag] time: false
Using docker registry: anoop
-- meep-virt-engine --
   + copy docker data
    Using: /home/ec2-user/AdvantEDGE/charts/meep-loc-serv --> /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-loc-serv
Cmd: [rm -r /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-loc-serv]
Result: rm: cannot remove '/home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-loc-serv': No such file or directory

Cmd: [cp -r /home/ec2-user/AdvantEDGE/charts/meep-loc-serv /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-loc-serv]
Result: cp: cannot create directory '/home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-loc-serv': No such file or directory

Error: Failed to copy data:  /home/ec2-user/AdvantEDGE/charts/meep-loc-serv  -->  /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-loc-serv
[ec2-user@ip-172-31-25-121 ~]$
```

I noticed these lines & they seem to be missing the file/directory specified in the LHS of the `COPY`. 

https://github.com/InterDigitalInc/AdvantEDGE/blob/master/go-apps/meep-virt-engine/Dockerfile#L25-L26

But - if I manually copy over the directories and files from `go-apps` to `bin` (_and_ edit / comment out the lines above) & then issue a direct `docker build` command like `for i in `ls` ; do docker build -t anoop/$i -f $i/Dockerfile . ; docker push anoop/$i ; done`, then I can get the images to build, but I had to comment out/edit the [lines](https://github.com/InterDigitalInc/AdvantEDGE/blob/master/go-apps/meep-virt-engine/Dockerfile#L25-L26
) above.

I'm sure I'm missing something quite basic.

**Comment by dilallkx** [comment]
Hi @anokun7 ,

Your meepctl configuration seems to have an error in the `meep.registry`. Please follow the following instructions from the runtim installation wiki page:
https://github.com/InterDigitalInc/AdvantEDGE/wiki/runtime-environment#step-7---configure-docker-registry

Here is what my config looks like:
```
========================================
meep.gitdir : /home/englab/dilallkx/AdvantEDGE
meep.registry : meep-docker-registry:30001
meep.workdir : /home/englab/.meep
node.ip : 10.190.114.150
========================================
```

The docker registry name is important and required to store the platform docker images correctly for deployment.
Please set the proper configuration and let me know if it resolves your issue.

**Comment by anokun7** [comment]
Thank you @dilallkx 

So I have tried using the default as well, which was `meep-docker-registry:30001`. But same issue. I wanted to share my images with my team & so I have my public docker hub name there. In any event, that should not have a bearing on building images per se, should it? I confirm that I get the same behavior using either value.

```
[ec2-user@ip-172-31-25-121 ~]$ meepctl dockerize meep-virt-engine -v
Using repo config file: /home/ec2-user/AdvantEDGE/.meepctl-repocfg.yaml
Using meepctl config file: /home/ec2-user/.meepctl.yaml
Dockerize called
[arg]  targets: [meep-virt-engine]
[flag] registry:
[flag] verbose: true
[flag] time: false
Using docker registry: meep-docker-registry:30001
-- meep-virt-engine --
   + copy docker data
    Using: /home/ec2-user/AdvantEDGE/charts/meep-gis-engine --> /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-gis-engine
Cmd: [rm -r /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-gis-engine]
Result: rm: cannot remove '/home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-gis-engine': No such file or directory

Cmd: [cp -r /home/ec2-user/AdvantEDGE/charts/meep-gis-engine /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-gis-engine]
Result: cp: cannot create directory '/home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-gis-engine': No such file or directory

Error: Failed to copy data:  /home/ec2-user/AdvantEDGE/charts/meep-gis-engine  -->  /home/ec2-user/AdvantEDGE/bin/meep-virt-engine/meep-gis-engine
```
```
[ec2-user@ip-172-31-25-121 ~]$ meepctl config
========================================
meep.gitdir : /home/ec2-user/AdvantEDGE
meep.registry : meep-docker-registry:30001
meep.workdir : /home/ec2-user/.meep
node.ip : 172.31.25.121
========================================
```
```
[ec2-user@ip-172-31-25-121 ~]$ cat /etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6

172.31.25.121 meep-docker-registry
[ec2-user@ip-172-31-25-121 ~]$
```

**Comment by dilallkx** [comment, marked as answer]
Hi @anokun7 ,

You are correct, you should be able to use a different docker registry and this does not seem to be the cause of the issue you are observing.

Your issue seems to be caused by a meep-virt-engine build failure. The folder `/home/ec2-user/AdvantEDGE/bin/meep-virt-engine` should be created when you build the meep-virt-engine and it should contain a copy of the Dockerfile and the meep-virt-engine binary. Please run the following command and provide the build logs to check for errors: `meepctl build meep-virt-engine -v`

**NOTE:** The microservice binaries must be successfully built using the `meepctl build` command before using the `meepctl dockerize` command to create the platform docker images.

Here is a link to the AdvantEDGE build wiki page:
https://github.com/InterDigitalInc/AdvantEDGE/wiki/build-advantedge#build-advantedge-micro-services

The build logs should give us more information about the error.

**Comment by anokun7** [comment]
Ah. That was it. 👍

<rant>
I was trying to get away from setting up a development/build environment. It seemed especially promising when I was able to download the `meepctl` binary without having to build it using a development environment. It seemed even more promising when I was able to run `meepctl deploy dep` without using docker at all.
 I think we can either host pre-built docker images or at least prepare a development environment within a docker container. It would be nice to have a deployment only option. It would be more friction-less that way, IMO.
</rant>

Thank you @dilallkx 🙏 - I was able to run `dockerize all` (after installing a bunch of stuff I have no idea what they do). This is on RHEL 8 by the way, using docker, containerd & k0s.

**Comment by dilallkx** [comment]
Hi @anokun7 ,

Glad to hear this fixed your issue!
Also glad to hear that you were able to create the docker images on RHEL 8 (note that we do not test this distribution).

WRT installation from binaries (to avoid dev environment setup) this is possible for major releases and is briefly described in the wiki procedure here: https://github.com/InterDigitalInc/AdvantEDGE/wiki#getting-started

You must extract the release binaries to the AdvantEDGE folder, which should create the /bin folder with the release binaries. The meepctl dockerize command should then work correctly. You mention that you were able to get the meepctl binary, the platform binaries should have been in the same archive. Please let me know if this is not the case.

**Note:** For release v1.6.1 hotfix, the [v1.6.0](https://github.com/InterDigitalInc/AdvantEDGE/releases/tag/v1.6.0) binaries should be used. I will update the v1.6.1 release notes accordingly.

Also, I agree that a containerized dev environment would be great... it is on our improvement roadmap!
Thanks again!
